### DISCUSSION TITLE
Can't communicate to panel

### DISCUSSION THREAD

**Discussion by JimO67** [aqualinkd/AqualinkD#209] [Q&A] [answered]
I believe all is working except for the connecting to RS 485 on the panel ,  I can connect to the web server.  Need to know if I am doing something fundamentally wrong or not with my wiring.  I have the RS 6 Combo.  I have two RS485 ports so I wired to the free 2nd one.  1st one has the PDA transceiver connected to it.  
The sudo -s eval 'stty raw -echo < /dev/ttyUSB0; cat -vte /dev/ttyUSB0' responds as shown int the instructions.

Left ID as 0x0a.  <---  Is this the issue?  Thinking this might be it.

  First just had the data+/- connected (how interpret the instructions on wiki) then tried adding green for ground.  Then tried black in place of green.  No difference.  I am using the recommended USB-RS485 product:  https://www.amazon.com/dp/B01HI9CMT2
They say Orange is data + and yellow data -.

![Screenshot 2023-05-05 125442](https://user-images.githubusercontent.com/90992628/236536824-23f23bfe-77f0-4d93-8f43-e0a22e9d5b18.jpg)

log screen shots:
![LOG2  2023-05-05 130241 (2)](https://user-images.githubusercontent.com/90992628/236537101-087e578d-e969-4911-a7ff-6e53a2534b82.jpg)
![LOG2  2023-05-05 130241](https://user-images.githubusercontent.com/90992628/236537158-6e151573-a698-4a0f-9a68-10eaad240a4f.jpg)

**Comment by ballle98** [comment]
"wait on probe" means it is not it's not seeing a probe message coming from the control panel.  The panel will send probe ever couple seconds.  you can connect both sets of wires to the same connector.  try serial loggin to see if RS485 is working https://github.com/sfeakes/AqualinkD/wiki#serial_log

**Reply by JimO67** [reply]
Ran the logger.  Thanks for helping.  I have not set the ID back to 0x0a but the logger came back with the following:
![image](https://user-images.githubusercontent.com/90992628/236549081-97f915bb-363c-4ada-97c1-c99c5a291a9a.png)

So do I use the ID listed there. 0X33?

Apparently not:
![image](https://user-images.githubusercontent.com/90992628/236549705-5e2975c4-ee8e-4cbc-a6a2-ccacd7560b46.png)

**Reply by ballle98** [reply]
This may be a PDA only pannel.  In which case you will have to use pda mode and set to 0x60-0x63.  My panel is PDA only and I have made several aditional PDA only enhancements on https://github.com/ballle98/AqualinkD

**Reply by JimO67** [reply]
I was just suspecting that.  I have no serial connected devices, like pumps/heaters.  Everything must hard wire/relay control.   Maybe heater set is just analog?  Anyhow,  trying 0x60.. now.

**Reply by JimO67** [reply]
Hmm

![image](https://user-images.githubusercontent.com/90992628/236551210-1078bc7d-5055-498d-b1b5-00a1d890903c.png)

Is this happening because i still have this:
![image](https://user-images.githubusercontent.com/90992628/236551524-5f2596c4-0fb4-4611-972f-138c0cc864f5.png)

Do I need to put PDA-6 Combo  ?

I'm in!  Thank you so much!  Frustration level was getting high!  
![image](https://user-images.githubusercontent.com/90992628/236554064-8c21c9ae-67f1-4015-947f-fbf88a254e88.png)

Need to map the waterfall motor.  I don't follow the settings on that.  I can trial and error.  If you can provide some hints, would appreciate.

**Reply by ballle98** [reply]
Use pda8 combo the logic around number of devices is not working 100%

**Reply by ballle98** [reply]
Look at https://github.com/sfeakes/AqualinkD/wiki#notes-on-aqualink-pda-only-non-rs-panels

**Reply by JimO67** [reply]
I was hoping this was not my case as I saw this on the chip.  Says PDA PS6
![Screenshot 2023-05-04 140109](https://user-images.githubusercontent.com/90992628/236556488-35c5b9d4-e717-470f-aaa5-9968ff0beef9.jpg)

**Comment by ballle98** [comment]
Honestly I have a PDA only panel and been very please using aqualinkd with it for many years.  I fixed many bugs over the years.  Most recently around SWG.

**Reply by JimO67** [reply]
Cool.  I just got my Pi Z into an enclosure mounted outside the pool controller box.  My RS-485 adapter seems to be shutting down after a couple of minutes.  Also, whenever I start up aqualinkd, the web server often fails to start (That was hapening b4 I put in the enclosure).    I try again a few times an it eventually starts.  Doing this over putty SSH. Once I figure out these problems, I need to make the Pi boot from the SD card.  And I assuem I make AqualinkD start on boot.  https://people.utm.my/shaharil/run-programs-on-your-raspberry-pi-at-startup/

All of my IOT projects have been with NodeMCU/ESP01 and Blynk and/or Arduino Cloud..   I think I will still work on making a version of this that runs on the nodemcu.  Then I can control from outside my LAN.  I know I could port foreward my router also.  I could probably make a login screen.    All this PI stuff seems very complex/expensive just to turn on/off a few things.  I believe we are just sending text, right?  So even the temperature is just a text value coming form the pool controller.  I don't need an adc.  

I already have a nodemcu monitoring my two pump currents (at the indoor breaker panel) .  It's probably  100' away from the pool controller. RS485 can be run 4000'.  Ideally I could just add a RS485 board there and run the serial line only to the pool controller.

**Reply by sfeakes** [reply]
You won't get AqualinkD running on NodeMCU, see https://github.com/sfeakes/AqualinkD/issues/210#issuecomment-1539200534
Even more with PDA mode that requires a lot of complex timing that won't happen on a NodeMCU/ESP device.  Not without a complete rewrite.
No is't not passing text, it's a binary protocol.  
No you don't need ADC, it's a RS485 protocol, you use ADC for electrical resistance communication.

As for how a Pi or ESP device connects to your lan, they both connect in an identical fashion. So I don't understand your comments around that.

If your Pi is booting off USB then you don't need to make it boot from SD as you stated. But obviously it needs to boot from something (USB or SD).  If you used the install script this it will start AqualinkD when the pi boots.

RS485 max length is 40ft without reducing data speed, over that you will be reducing the communication speed, you can go to 4000ft, but with severely limiting the data rate.

As for the the web server not starting etc, it really sounds like you need to read the wiki and set the appropiate config settings. Have you tried serial_logger to get the output from your panel?

**Reply by JimO67** [reply]
1. I'm talking about a rewrite specific for the NodeMCU.  There is example code out there for using the NodeMCU with an RS485.  It's not like I would be the first.  I figured I could then just leverage what is in the AqualinkD code to handshake/pass information.  Not a problem to wait for some signal, delay, send something, delay, read something.  Just need to know the protocol.

2. I am about to go to pull the logs, but I don't think the error message going to be anything other than the same being echoed at runtime in the terminal window.  Before I put the pi in an enclosure I saw the web server failures I never had the RS-485 Adapter drop out/appear dead.  I see no lights flashing on it.  I am concerned about heat in the water tight plastic enclosure.  I doubt this is it what is happening as it happens pretty soon.  Within a minute.  The pi does not seem to have a problem.  Still connect to it over putty.   My other concern is I added a voltage converter to get the power from the RS485 port.  I saw about 10.6V on it and adjusted the converter to provide a 5V output.  I did that on the bench.  Power wiring is not shielded.  Was thinking maybe I was getting noise on the power line.  If it was noise on the line, I don't think the adapter would just drop out and appear dead.  

3. I want to be able to access my pool controller over the internet.  I can do that if I set up my router for forwarding to the Pi, what security concerns does that bring?  I have done this before with username and password was available (a camera).  Why doesn't AqualinkD just have a login and password?  Is that not enough? To log in to the pi via telnet/ssh you have to get past a login.

4. My NodeMCU/Arduino projects use an IOT service like Blynk or Arduino Cloud service.  They cover the security aspect.  My microcontrollers get a long ID and password.  My perception is they are pretty safe.
 
5. I interpreted the 1st paragraph of the WIKI to say I need to do the read only root step.  "All Raspberry PI's (except the Pi 3) are inherently unstable devices to be running 24/7/365 in default Linux configurations. This is due to the way they use CF card, a power outage will generally cause CF card corruption."  I am running a pi zw with Raspbian.   I boot from the microSD card. The external USB port is used by the RS485 adapter only,  I don't need to do the read only root?

Thanks for the help.  If I can get past the security concerns, I don't really have a motivation to implement the NodeMCU solution.

**Reply by ballle98** [reply]
@JimO67 I don't know if your pool controller is on the side of the house or garage.  If it is, it's much kinder to the RPi to put it somewhere that is climate controlled and run a wire to the controll panel.  Inside the pannel there is little air flow and on a hot day you could exceed the 70 C operating temp or on a cold day drop below 0 C.

**Reply by JimO67** [reply]
Update:
1.  Via top command  I see the aqualinkd is already running on start up.  It took longer than i expected;but I was able to reach Aqualinkd on chrome with my phone and  it has been working fine.  The rs485 adapter trans/rec lights still flashing.  Was I opening up additional instances by rerunning the program?  I closed the enclosure door and still going,  I did also check my wires and tighten one.

I found: "There is no security, and .htaccess files will be ignored. You could add very very basic security to AqualinkD, but it would require code changes, and quite frankly would never be all that great / secure. This is a limitation of a library I use called mongoose. If you do want security, then I would strong suggest installing a real web server, like Apache or nginx and using that to proxy the requests to AqualinkD, and put the security in the real web server. There is no way you can beat/compete with the security those services will offer."

So help on 5 is all I need now.  Do I need to do the root installation or not?
 


For benefit of others, my setup:
![image](https://github.com/sfeakes/AqualinkD/assets/90992628/4c15024b-0aba-49b6-a40f-2951328b72ee)
![image](https://github.com/sfeakes/AqualinkD/assets/90992628/e6f23266-1eb4-40da-8e86-cdbd0d19bd77)
Will see if it handles the heat in TX.

**Reply by sfeakes** [reply, marked as answer]
ReadOnly Root, is probably something you should not tackle yet.  The newest RaspberryOS releases overcome the hardware limitations on the Pi very well and corrupt SD cards as not as common as they once were. ReadOnly Root is to overcome the issue with corrupt CF cards. 

Their are some pointers to securing AqualinkD in the other thread here.
https://github.com/sfeakes/AqualinkD/issues/143
